node/src/reactor/validator/tests.rs

```diff
@@ -241,23 +214,28 @@ async fn run_equivocator_network() {
         .await
         .expect("network initialization failed");
 
+    // With an auction delay of 1 and an unbonding delay of 3, there are always 2 bonded active
+    // eras: the current one and the one before. The era supervisor keeps three active eras in
+    // memory — the oldest one without units, just for validating evidence if needed. So if we run
+    // this for five eras, the test can catch issues with unbonding and dropping obsolete eras.
```

Comment:
Comment in era supervisor says:
```
/// A node keeps `2 * bonded_eras` past eras around, because the oldest bonded era could still
/// receive blocks that refer to `bonded_eras` before that.
```
and
```
/// This map always contains exactly `2 * bonded_eras + 1` entries, with the last one being the
/// current one.
active_eras: HashMap<EraId, Era<I>>,
```
and
```
let bonded_eras: u64 = protocol_config.unbonding_delay - protocol_config.auction_delay;
```
so with `auction_delay=1` and `unbonding_delay=3, we have `bonded_eras=2` which means we keep 5 (`2 * bonded_eras + 1`) in `active_eras`?

Reply:
Ugh, more off-by-one errors! :weary: 
Good catch!
~Maybe it's the era supervisor (or just its comment) that's wrong here, though, and keeping more eras in memory than necessary?~ Each time this comes up I have to start from the beginning…

Let `A` be the auction delay and `U` the unbonding delay. So at the end of era `N`, we select the validators for era `N + A + 1` — there are `A` full eras between the auction and the beginning of the era run by the highest bidders.

If you are still a validator in era `N` that means you haven't made an unbonding request during era `N - A - 1`. So your stake will not be withdrawn at the end of era `N - A - 1 + U`. The earliest possible withdrawal is at the end of era `N - A + U`. So: During era `N - A + U`, all validators from eras back to and including `N` are still bonded. Let `B = U - A`. So:
* In era `N`, validators from era `N - B` are still bonded, so we should slash them if we can. That means that era `N - B` should ban validators that were slashed in `N - B` or later, and slash validators for which new evidence in era `N - B` or later occurs. So: As long as we keep track of the _units_ in `N`, we have to keep track of _evidence_ in `N - B`.
* It also means that it might be worth trying to _detect_ new equivocations back to era `N - B` (which means keeping track of its units) until the moment the switch block in era `N` is proposed — that's the last chance to slash them. On the other hand we shouldn't track `N - B`'s units longer than that, because after the switch block has been proposed, the validators can freely equivocate and spam us. (But [CEP 36](https://github.com/CasperLabs/ceps/blob/master/text/0036-consensus-memory-usage.md) will drop them much earlier anyway.)

 I believe that's what the era supervisor is currently doing, even though it calls `B = U - A` `bonded_eras` instead of `bonded_past_eras`. (And it only drops stuff when the switch block is finalized and executed, not when it's proposed.) :confused: 
(E.g. that's why we have the `+ 1` in [this line](https://github.com/CasperLabs/casper-node/blob/cb1d20ad1ea6e245cd8237f9406885a1e785c669/node/src/components/consensus/era_supervisor.rs#L371).)

Anyway, you're right, my comment is not in line with the era supervisor; I'll update it:
* `B` is actually `2` here.
* So at the beginning of era 5 we clear the units from era 2…
* and _drop_ era 0.

Reply:
Done in https://github.com/CasperLabs/casper-node/pull/1052/commits/f2d7ab490ca30be910050f3377f1eddc9df5faf6.